### PR TITLE
feat: implement `/profile/{id}/points` endpoint

### PR DESCRIPTION
**Pull request by elijaholmos**
- implements the backend for https://github.com/leyline-gg/pog-monorepo/issues/212
- directly interfaces with the ledger
- if invalid profile ID is provided, error will be returned from leger and handled in `src/services/proof-of-good-smart-contract.service.ts`
  - it will be converted to an `ENTITY_NOT_FOUND` error, causing Loopback to set response status to 404

### CODE REVIEW

File: src/services/proof-of-good-smart-contract.service.ts [hunk 1]
```diff
@@ -167,6 +167,34 @@ export class ProofOfGoodSmartContractService {
...
+    try {
+      const res = await this.contract.profileByUserId(userId);
+      const [
+        profileId,
+        walletAddresses,
+        balance,
+        totalGood,
+        categories,
+        entries,
+      ] = res;
```
Comment (kharaone) [inline src/services/proof-of-good-smart-contract.service.ts]:
```suggestion
      const [
        profileId,
        walletAddresses,
        balance,
        totalGood,
        categories,
        entries,
      ] = await this.contract.profileByUserId(userId);
```
we don't use res elsewhere :)

File: src/services/proof-of-good-smart-contract.service.ts [hunk 2]
```diff
@@ -167,6 +167,34 @@ export class ProofOfGoodSmartContractService {
...
+        totalGood: totalGood.toNumber(),
+      };
+    } catch (error) {
+      throw error?.reason === 'profile not found'
+        ? {
+            code: 'ENTITY_NOT_FOUND',
+            entityName: 'pogprofiles',
+            entityId: userId,
+          }
+        : error;
```
Comment (kharaone) [inline src/services/proof-of-good-smart-contract.service.ts:204]: feels feature envy, we should have a better way to return know error type.
If works for now, but we should definitely have a better mechanism (custom error, factory, ....) to get those errors created without forging them ourselves at that level of implementation

File: src/services/proof-of-good-smart-contract.service.ts [hunk 3]
```diff
@@ -167,6 +167,34 @@ export class ProofOfGoodSmartContractService {
       return this.contract.setCap(activityId, oracleId, duration, points);
     });
   }
+
+  async getUserGoodPoints(userId: string) {
+    try {
+      const res = await this.contract.profileByUserId(userId);
```
Comment (kharaone) [inline src/services/proof-of-good-smart-contract.service.ts]: Given that I'm requesting at least one change, could we encapsulate this within the ethers poll try method to make it a bit more resilient ?

**Review by kharaone — CHANGES_REQUESTED** [review]
Looks good !

a bit more resilience and we're good :)

**Comment by kharaone** [comment]
@elijaholmos

File: src/services/proof-of-good-smart-contract.service.ts (hunk 1 above)
Comment (elijaholmos) [reply src/services/proof-of-good-smart-contract.service.ts]: good catch

File: src/services/proof-of-good-smart-contract.service.ts (hunk 2 above)
Comment (elijaholmos) [reply src/services/proof-of-good-smart-contract.service.ts:204]: I agree. a lot of it depends on how Loopback parses errors to set different response status codes, which will require some research, at least with the limited knowledge I have. 500 appears to be the default response status for most errors.

File: src/services/proof-of-good-smart-contract.service.ts (hunk 3 above)
Comment (elijaholmos) [reply src/services/proof-of-good-smart-contract.service.ts]: yes, but give me a few hours to get home and implement the change :)

Comment (elijaholmos) [reply src/services/proof-of-good-smart-contract.service.ts]: @kharaone take a look at 048320fff9717861d877dc70897b6c0796f78c97 and let me know if my implementation should be modified... I'm not sure if the `eventName`/`events.find(...)` logic needs to be sustained for a read

**Review by kharaone — APPROVED** [review]
LGTM :)

- I added the _ so that the linter ignores the unused variables
- I added a log of the result

**Comment by elijaholmos** [comment]
thank you!

**Pull request merged**